src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -569,6 +592,129 @@ int QgsPointCloudLayerRenderer::renderNodesSorted( const QVector<IndexedPointClo
   return blockCount;
 }
 
+inline bool isEdgeTooLong( const QPointF &p1, const QPointF &p2, float length )
+{
+  QPointF p = p1 - p2;
+  return p.x() * p.x() + p.y() * p.y() > length;
+}
+
+static void renderTriangle( QImage &img, QPointF *pts, QRgb c0, QRgb c1, QRgb c2, float horizontalFilter, float *elev, QgsElevationMap *elevationMap )
+{
+  if ( horizontalFilter > 0 )
+  {
+    float filterThreshold2 = horizontalFilter * horizontalFilter;
+    if ( isEdgeTooLong( pts[0], pts[1], filterThreshold2 ) ||
+         isEdgeTooLong( pts[1], pts[2], filterThreshold2 ) ||
+         isEdgeTooLong( pts[2], pts[0], filterThreshold2 ) )
+      return;
+  }
+
+  QgsRectangle screenBBox = QgsMeshLayerUtils::triangleBoundingBox( pts[0], pts[1], pts[2] );
+
+  QSize outputSize = img.size();
+
+  int topLim = std::max( int( screenBBox.yMinimum() ), 0 );
+  int bottomLim = std::min( int( screenBBox.yMaximum() ), outputSize.height() - 1 );
+  int leftLim = std::max( int( screenBBox.xMinimum() ), 0 );
+  int rightLim = std::min( int( screenBBox.xMaximum() ), outputSize.width() - 1 );
+
+  int red0 = qRed( c0 ), green0 = qGreen( c0 ), blue0 = qBlue( c0 );
+  int red1 = qRed( c1 ), green1 = qGreen( c1 ), blue1 = qBlue( c1 );
+  int red2 = qRed( c2 ), green2 = qGreen( c2 ), blue2 = qBlue( c2 );
```

Comment:
There's an opportunity here for an optimisation -- we could check if c0 == c1 == c2 (and not elevation map), and if so then just use the QPainter renderPolygon method to draw a filled solid color triangle. 

It's a situation that WILL arise frequently, eg for classified point clouds.

Reply:
I have tried this, but unfortunately it leaves ugly artifacts at the edges between barycentric-based and QPainter-based triangles - even when I use 1px solid pen in addition to the brush fill for QPainter, and I would prefer not to use thicker pens. After this initial work, I would like to do a bit of profiling, so if possible I would prefer to leave optimisation for later (moreover, for attribute by ramp or RGB, the QPainter solid triangle path would not get triggered often)

![image](https://github.com/qgis/QGIS/assets/193367/71435f56-201f-452b-8eb7-0565cf782846)

Reply:
Ok, fair enough!